### PR TITLE
codership/mariadb-wsrep#27 Galera cache encryption

### PR DESCRIPTION
**Pull request by mkaruza**
* Created interface class for encryption support
* Implemented function for setting enc key to provider, callback function for encryption/decryption

### CODE REVIEW

File: include/wsrep/encryption_service.hpp [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+        /**
+         * Return encryption key.
+         */
+        const std::vector<unsigned char>& get_encryption_key() const
+        { 
+            return encryption_key_;
+        }
+
+    protected:
+        std::vector<unsigned char> encryption_key_;
```
Comment (temeo) [inline include/wsrep/encryption_service.hpp]: Why we need to store encryption key here? Ideally the service classes should be stateless, i.e. not have member variables.

If the `encryption_key_` would be part of the `server_state`, the getters and setters should be unnecessary here, making this interface pure.

File: include/wsrep/server_state.hpp [hunk 2]
```diff
@@ -180,8 +181,10 @@ namespace wsrep
 
         virtual ~server_state();
 
-
+        wsrep::encryption_service* encryption_service()
+        { return encryption_service_; }
```
Comment (temeo) [inline include/wsrep/server_state.hpp]: Why we need a public access to `encryption_service`? It should be used only internally by `server_state`. Is is because of encryption key getters and setters are part of the service interface?

File: src/wsrep_provider_v26.cpp [hunk 3]
```diff
@@ -568,7 +597,7 @@ wsrep::wsrep_provider_v26::wsrep_provider_v26(
     init_args.connected_cb = &connected_cb;
     init_args.view_cb = &view_cb;
     init_args.sst_request_cb = &sst_request_cb;
-    init_args.encrypt_cb = &encrypt_cb;
+    init_args.encrypt_cb = encrypt_cb;
```
Comment (temeo) [inline src/wsrep_provider_v26.cpp]: To my understanding, we should set `init_args.encrypt_cb = 0` here if the encryption service is not provided.

File: dbsim/db_server_state.hpp [hunk 4]
```diff
@@ -40,6 +40,7 @@ namespace db
                 mutex_,
                 cond_,
                 server_service,
+                NULL,
```
Comment (temeo) [inline dbsim/db_server_state.hpp]: Could use `nullptr` here (dbsim is compiled with -std=c++14).

**Comment by temeo** [comment]
As a general direction, I would remove the `encryption_key_` member for `encryption_service` interface and keep it a member of `server_state`. Also the key should be set via `server_state`. This would eliminate the key settter and getter from `encryption_service` interface and simplify the code significantly.

**Comment by temeo** [comment]
And even further, is there any reason to store `encryption_key_` in either `encryption_service` or `server_state`? Is there any use case where it needs to be stored instead of just passing the buffer to the provider?

File: include/wsrep/server_service.hpp [hunk 5]
```diff
@@ -226,7 +226,6 @@ namespace wsrep
                               const wsrep::gtid& gtid,
                               bool bypass) = 0;
 
-
```
Comment (ayurchen) [inline include/wsrep/server_service.hpp]: and changes to this file seem to be irrelevant )

File: include/wsrep/encryption_service.hpp [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+
+        /**
+         * Set encryption key
+         * 
+         * @param provider WSREP provider
+         * @param key Encryption key
+         * 
+         * @return Zero on success, non-zero on failure.
+         */
+        virtual int set_encryption_key(wsrep::provider& provider,
```
Comment (ayurchen) [inline include/wsrep/encryption_service.hpp]: `set_encryption_key()` does not seem to belong to encryption service interface at all as it is ultimately a call to `server_state`.
So while `provider::enc_set_key()` may have to be called directly *from the encryption service implementation*, there seem to be no reason for it to be part of the *wsrep-lib* encryption service interface. It is not something that provider would call.

File: include/wsrep/encryption_service.hpp (hunk 1 above)
Comment (mkaruza) [reply include/wsrep/encryption_service.hpp]: Refactored. Set/get key & key container is moved to server_state.

File: include/wsrep/server_state.hpp (hunk 2 above)
Comment (mkaruza) [reply include/wsrep/server_state.hpp]: Encrypt_cb needs to use do_crypt function.

File: src/wsrep_provider_v26.cpp (hunk 3 above)
Comment (mkaruza) [reply src/wsrep_provider_v26.cpp]: Callback will be set if encryption service is provided & enabled.

File: dbsim/db_server_state.hpp (hunk 4 above)
Comment (mkaruza) [reply dbsim/db_server_state.hpp]: Changed.

File: src/wsrep_provider_v26.cpp [hunk 7]
```diff
@@ -582,6 +606,19 @@ wsrep::wsrep_provider_v26::wsrep_provider_v26(
...
+    {
+        const std::vector<unsigned char>& key = server_state.get_encryption_key();
+        if (key.size())
+        {
+            wsrep::const_buffer const_key(key.data(), key.size());
+            if(enc_set_key(const_key))
+            {
+                throw wsrep::runtime_error("Failed to set encryption key");
+            }
+        }
```
Comment (temeo) [inline src/wsrep_provider_v26.cpp]: Should the case `encryption_enabled` and no encryption key set be considered an error here?

File: src/wsrep_provider_v26.cpp [hunk 8]
```diff
@@ -421,14 +422,35 @@ namespace
...
+                   const wsrep_buf_t*    input,
+                   void*                 output,
+                   wsrep_enc_direction_t direction,
+                   bool                  last)
     {
-        return 0;
+        assert(app_ctx);
+        wsrep::server_state& server_state(
+            *static_cast<wsrep::server_state*>(app_ctx));
+
```
Comment (temeo) [inline src/wsrep_provider_v26.cpp]:
```suggestion
assert(server_state.encryption_service());
if (server_state.encryption_service() == 0)
{
   wsrep::log_error() << "Encryption service not defined in encrypt_cb()";
   return -1;
}
```

File: include/wsrep/encryption_service.hpp [hunk 9]
```diff
@@ -0,0 +1,69 @@
...
+                             void*                 output,
+                             bool                  encrypt,
+                             bool                  last) = 0;
+
+        /**
+         * Is encryption enabled on server.
+         * 
+         * @return 1 if encryption is enabled. 0 otherwise
+         */
+        virtual int encryption_enabled() = 0;
```
[the remark this answers is not in the document]
Comment (mkaruza) [reply include/wsrep/encryption_service.hpp]: I'll change.

File: src/wsrep_provider_v26.cpp (hunk 7 above)
Comment (mkaruza) [reply src/wsrep_provider_v26.cpp]: Good question, but i think not. Maybe key is not set at that point and will be provided later. @ayurchen thoughts ?

File: src/wsrep_provider_v26.cpp (hunk 8 above)
Comment (mkaruza) [reply src/wsrep_provider_v26.cpp]: Encryption_cb is already set only if there is service provided & enabled. But this will not hurt to add for extra protection.

File: src/wsrep_provider_v26.cpp (hunk 7 above)
Comment (ayurchen) [reply src/wsrep_provider_v26.cpp]: Well, 0-sized key is technically a key. But seriously, wsrep API allows setting a key at any time between provider initialization and `connect()` call. (The reason for it was to have a single way to set the key)
So here, I think it boils down to a decision:
 * freak out right away OR
 * freak out before `connect()` call if the key is still 0

The upside of the second approach is greater application initialization flexibility. The downside is that we may freak out much later in the process. And while this does not change wsrep-lib signatures, we must decide on it now as it defines the order of initialization calls.

File: include/wsrep/encryption_service.hpp [hunk 10]
```diff
@@ -0,0 +1,69 @@
...
+ * You should have received a copy of the GNU General Public License
+ * along with wsrep-lib.  If not, see <https://www.gnu.org/licenses/>.
+ */
+
+#ifndef WSREP_ENCRYPTION_SERVICE_HPP
+#define WSREP_ENCRYPTION_SERVICE_HPP
+
+#include "provider.hpp"
+
+#include <vector>
```
Comment (ayurchen) [inline include/wsrep/encryption_service.hpp]: do we need this include any more?
